package de.hhu.bsinfo.panama.upcall;

import jdk.incubator.foreign.Addressable;

public class UpcallHelper {
    /**
     * This method would normally be generated by jextract. However, the generated method calls mh$.invokeExact()
     * inside a try-Block, which throws an AssertionError, once it catches a Throwable.
     *
     * Since we want to make sure, we can catch a RuntimeException inside the main-method,
     * we implement this modified version of perform_upcall(), which does not call mh$.invokeExact() inside a try-Block,
     * but lets the Throwable through to the caller.
     */
    public static void perform_upcall ( Addressable upcall) throws Throwable {
        var mh$ = RuntimeHelper.requireNonNull(constants$0.perform_upcall$MH, "perform_upcall");
        mh$.invokeExact(upcall);
    }
}
